include/proteus/Utils.h

```diff
@@ -21,12 +22,18 @@
 #include "proteus/TimeTracing.hpp"
 
 template <typename T> void saveToFile(llvm::StringRef Filepath, T &&Data) {
+  // Write to temp file first, then rename
+  // so readers never see a partial file.
+  std::string TempPath = Filepath.str() + ".tmp";
+
   std::error_code EC;
-  llvm::raw_fd_ostream Out(Filepath, EC);
+  llvm::raw_fd_ostream Out(TempPath, EC);
   if (EC)
-    PROTEUS_FATAL_ERROR("Cannot open file" + Filepath);
+    PROTEUS_FATAL_ERROR("Cannot open file " + TempPath);
   Out << Data;
   Out.close();
+
+  std::filesystem::rename(TempPath, Filepath.str());
 }
```

Comment:
Create a `saveToFileAtomic` besides the existing `saveToFile` to do (atomically) rename only when needed